### PR TITLE
gfx: Fix border-radius panic when a corner has 0px and >0px borders

### LINKED ISSUE #7700: border-radius > 0 between horizontal border with width > 0 and vertical border with width == 0 causes panic in intersect_ellipse_linein combination with
The following code causes the crash:

```html
<html>
<head>
</head>
<body>
<div style="border-top: 30px solid black; border-radius: 15px"></div>
<div style="border-right: 30px solid black; border-radius: 15px; height: 30px"></div>
<div style="border-bottom: 30px solid black; border-radius: 15px"></div>
<div style="border-left: 30px solid black; border-radius: 15px; height: 30px"></div>
</body>
```

The issue is caused by `draw_corner` not dealing correctly with 0-width borders [here](https://github.com/servo/servo/blob/master/components/gfx/paint_context.rs#L675-L678).

Replacing the `if inner_border.x >= outer_border.x {` with the following conditions fixes the crash, but I'm not sure it's the right solution:

```rust
if inner_border.x == outer_border.x {
  0f32
} else if inner_border.x > outer_border.x {
...
```

### PR DESCRIPTION
**Pull request by bjwbell**
When one border is 0px and the other is >0px then the border corner
drawing code panics when computing the values to use in drawing the
border corner arcs.

This fixes that bug and makes the `draw_corner` function more robust
by explicitly passing an enum, `BorderCorner`, naming which corner is
being drawn e.g. `BorderCorner::TL`.

Add a ref test,
`border_radius_zero_sizes_a.html/border_radius_zero_sizes_ref.html`.

Fixes https://github.com/servo/servo/issues/7700.

r? @pcwalton or @mbrubeck

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7703)

### CODE REVIEW

File: components/gfx/paint_context.rs [hunk 1]
```diff
@@ -83,6 +91,23 @@ struct Line {
     start: Point2D<f32>,
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Add a comment to this struct stating the invariant that you're now upholding in the constructor

File: components/gfx/paint_context.rs [hunk 2]
```diff
@@ -83,6 +91,23 @@ struct Line {
...
     end: Point2D<f32>,
 }
+impl Line {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: uber-nit: newline before `impl`

File: components/gfx/paint_context.rs [hunk 3]
```diff
@@ -394,12 +419,34 @@ impl<'a> PaintContext<'a> {
         (Some(x1), Some(x2))
     }
 
-    fn intersect_ellipse_line(e: Ellipse, l: Line) -> (Option<Point2D<f32>>, Option<Point2D<f32>>) {
-        debug_assert!(l.end.x - l.start.x > f32::EPSILON, "Error line segment end.x > start.x!!");
-        // shift the origin to center of the ellipse.
-        let line = Line { start: l.start - e.origin,
-                          end: l.end - e.origin };
+    fn intersect_ellipse_line(_e: Ellipse, _l: Line) -> (Option<Point2D<f32>>,
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: You can just write `fn intersect_ellipse_line(mut e: Ellipse, mut line: Line)`

File: components/gfx/paint_context.rs [hunk 4]
```diff
@@ -394,12 +419,34 @@ impl<'a> PaintContext<'a> {
...
+            if clockwise {
+                // rotate clockwise by 90 degress
+                Point2D::new(point.y, -point.x)
+            } else {
+                // rotate counter clockwise by 90 degress
+                Point2D::new(-point.y, point.x)
+            }
+        }
+
+        // if line height is greater than it's width then rotate the axises by 90 degrees,
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: typo: "its"

File: components/gfx/paint_context.rs [hunk 5]
```diff
@@ -394,12 +419,34 @@ impl<'a> PaintContext<'a> {
...
-        // shift the origin to center of the ellipse.
-        let line = Line { start: l.start - e.origin,
-                          end: l.end - e.origin };
+    fn intersect_ellipse_line(_e: Ellipse, _l: Line) -> (Option<Point2D<f32>>,
+                                                         Option<Point2D<f32>>) {
+        let mut line = _l;
+        let mut e = _e;
+        let mut rotated_axises = false;
 
+        fn rotate_axises(point: Point2D<f32>, clockwise: bool) -> Point2D<f32> {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: `axes`?

File: components/gfx/paint_context.rs [hunk 6]
```diff
@@ -394,12 +419,34 @@ impl<'a> PaintContext<'a> {
...
+            if clockwise {
+                // rotate clockwise by 90 degress
+                Point2D::new(point.y, -point.x)
+            } else {
+                // rotate counter clockwise by 90 degress
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: typo: "degrees" (also in comment above)

File: components/gfx/paint_context.rs (hunk 5 above)
Comment (bjwbell) [inline components/gfx/paint_context.rs]: Yes axes is right. I looked up the wrong plural, http://www.thefreedictionary.com/axises. Axises is the plural form of an animal.

File: components/gfx/paint_context.rs (hunk 1 above)
Comment (bjwbell) [inline components/gfx/paint_context.rs]: Added

**Comment by bjwbell** [comment]
@pcwalton thanks for the fast review. Nits should be fixed.

File: components/gfx/paint_context.rs [hunk 7]
```diff
@@ -424,14 +471,24 @@ impl<'a> PaintContext<'a> {
                 if x0 > x1 {
                     mem::swap(&mut p0, &mut p1);
                 }
+                if rotated_axes {
+                    p0 = rotate_axes(p0, false);
+                    p1 = rotate_axes(p1, false);
+                }
                 (Some(p0), Some(p1))
             },
             (Some(x0), None) => {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Is it just me or can this be written `(Some(x0), None) | (None, Some(x0)) => { ... }`?

Comment (bjwbell) [inline components/gfx/paint_context.rs]: nice catch, yes it can.

**Comment by nox** [comment]
Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[components/gfx/paint_context.rs, line 74 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7703#-Jzo-WnXWN0J0R3GyN9n)** ([raw file](https://github.com/servo/servo/blob/071acf476b92bedf6ef90e653ebfa483d4d6ee74/components/gfx/paint_context.rs#L74)):</sup>
Could these not be abbreviated?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7703)

**Comment by bjwbell** [comment]
<sup>**[components/gfx/paint_context.rs, line 74 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7703#-Jzo-WnXWN0J0R3GyN9n)** ([raw file](https://github.com/servo/servo/blob/071acf476b92bedf6ef90e653ebfa483d4d6ee74/components/gfx/paint_context.rs#L74)):</sup>
I originally had them unabbreviated but for where they're used it seemed overly verbose. 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7703)

**Comment by Ms2ger** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/gfx/paint_context.rs, line 74 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7703#-Jzo-WnXWN0J0R3GyN9n)** ([raw file](https://github.com/servo/servo/blob/071acf476b92bedf6ef90e653ebfa483d4d6ee74/components/gfx/paint_context.rs#L74)):</sup>
My default stance is "no abbreviations"

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7703)

**Comment by bjwbell** [comment]
<sup>**[components/gfx/paint_context.rs, line 74 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7703#-Jzo-WnXWN0J0R3GyN9n)** ([raw file](https://github.com/servo/servo/blob/071acf476b92bedf6ef90e653ebfa483d4d6ee74/components/gfx/paint_context.rs#L74)):</sup>
I changed it back to the non-abbreviated version.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7703)

**Pull request merged**